host/src/server/api/v1/proof.rs

```diff
@@ -51,16 +52,48 @@ fn set_cached_input(
     };
 
     let path = get_input_path(dir, block_number, network);
-
-    if path.exists() {
-        return Ok(());
-    }
+    info!("caching input for {path:?}");
 
     let file = File::create(&path).map_err(<std::io::Error as Into<HostError>>::into)?;
+    bincode::serialize_into(file, input).map_err(|e| HostError::Anyhow(e.into()))
+}
 
-    info!("caching input for {path:?}");
+async fn validate_cache_input(
+    cached_input: Option<GuestInput>,
+    provider: &RpcBlockDataProvider,
+) -> HostResult<GuestInput> {
+    if let Some(cache_input) = cached_input {
+        debug!("Using cached input");
+        let blocks = provider
+            .get_blocks(&[(cache_input.block_number, false)])
+            .await?;
+        let block = blocks
+            .first()
+            .ok_or_else(|| RaikoError::RPC("No block data for the requested block".to_owned()))?;
 
-    bincode::serialize_into(file, input).map_err(|e| HostError::Anyhow(e.into()))
+        // double check if cache is valid for now
+        debug!(
+            "cache_input.block_hash_reference = {:?}",
+            cache_input.block_hash_reference
+        );
+        debug!(
+            "block.header.hash.unwrap_or(to_header(&block.header).hash()) = {:?}",
+            block.header.hash.unwrap_or(to_header(&block.header).hash())
+        );
+        if cache_input.block_hash_reference
+            == block.header.hash.unwrap_or(to_header(&block.header).hash())
+        {
+            return Ok(cache_input);
+        } else {
+            Err(HostError::InvalidRequestConfig(
+                "Cached input is not valid".to_owned(),
+            ))
+        }
+    } else {
+        Err(HostError::InvalidRequestConfig(
+            "Cached input is not enabled".to_owned(),
+        ))
+    }
```

Comment:
Maybe I'm nitpicking, but I'd write this function as follows:
```rust
    let Some(cache_input) = cached_input else {
        return Err(HostError::InvalidRequestConfig(
            "Cached input is not enabled".to_owned(),
        ));
    };

    debug!("Using cached input");

    let blocks = provider
        .get_blocks(&[(cache_input.block_number, false)])
        .await?;
    let block = blocks
        .first()
        .ok_or_else(|| RaikoError::RPC("No block data for the requested block".to_owned()))?;

    // double check if cache is valid for now
    debug!(
        "cache_input.block_hash_reference = {:?}",
        cache_input.block_hash_reference
    );

    let header_hash = block.header.hash.unwrap_or(to_header(&block.header).hash());
    debug!("block.header.hash.unwrap_or(to_header(&block.header).hash()) = {header_hash:?}");

    if cache_input.block_hash_reference == header_hash {
        Ok(cache_input)
    } else {
        Err(HostError::InvalidRequestConfig(
            "Cached input is not valid".to_owned(),
        ))
    }
```

Reply:
Yes, can be, updated!